<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="New skills | Portfolio Pedro de Bos" />
<meta property="og:type" content="book" />







<meta name="description" content="New skills | Portfolio Pedro de Bos">

<title>New skills | Portfolio Pedro de Bos</title>

<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="libs/navigation-1.1/tabsets.js"></script>
<script src="libs/htmlwidgets-1.5.4/htmlwidgets.js"></script>
<link href="libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet" />
<script src="libs/datatables-binding-0.23/datatables.js"></script>
<link href="libs/dt-core-1.11.3/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="libs/dt-core-1.11.3/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="libs/dt-core-1.11.3/js/jquery.dataTables.min.js"></script>
<link href="libs/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet" />
<script src="libs/crosstalk-1.2.0/js/crosstalk.min.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
/* show arrow before summary tag as in bootstrap
TODO: remove if boostrap in updated in html_document (rmarkdown#1485) */
details > summary {
  display: list-item;
  cursor: pointer;
}
</style>
</head>

<body>

<div class="container-fluid main-container">


<div class="row">
<div class="col-sm-12">
<!--bookdown:toc:end-->
<!--bookdown:toc:start-->
</div>
</div>
<div class="row">
<div class="col-sm-12">
<div id="new-skills" class="section level1">
<h1>New skills</h1>
<body id="start">
<div class="topnav">
<p><a href='index.html#Frontpage'>Frontpage</a>
<a href='data-visualisation.html#data-visualisation'>Data visualisation</a>
<a href='parametized-data-germany.html#parametized-data'>Parametizing data</a>
<a href='directory-structure.html#directory-structure'>Directory structure</a>
<a href='creating-a-r-package.html#creating-a-r-package'>R-package</a>
<a href='sql.html#SQL'>SQL</a>
<a href='bibliography-using-zotero.html#Bibliography using Zotero'>Zotero</a>
<a href='open-reproductibility-analysis.html#open-reproductibility-analysis'>Reproductibility</a>
<a href='future-endeavours.html#future-endeavours'>Future endeavours</a>
<a href='new-skills.html#new-skills'> New skills (Machine learning)</a>
<a href='cv.html#cv'>CV</a>
<a href='bibliography.html#bibliography'>Bibliography</a></p>
</div>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb67-1"><a href="new-skills.html#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb67-2"><a href="new-skills.html#cb67-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb67-3"><a href="new-skills.html#cb67-3" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span>opts_chunk<span class="sc">$</span><span class="fu">set</span>(<span class="at">eval=</span><span class="cn">FALSE</span>)</span></code></pre></div>
<p>For my free Space, I will study <a href="https://towardsdatascience.com/recognition-and-counting-of-microorganisms-on-petri-dishes-53a499283dc8">this site</a>. in order to find information about using bacterial identification using machine learning.</p>
<p>In this article, there is a link too the site <a href="https://agar.neurosys.com/">agar.neurosys.com</a> which contains a dataset containing picture of agarplates in different conditions</p>
<p>Conclusion based on the site: this is <em>way</em> too complicated to start out with. Trying to find a simpler introduction to machine learning, specifically in R</p>
<p>Trying <a href="https://www.datacamp.com/tutorial/machine-learning-in-r">this link</a>, mayhaps this’ll give a better introduction.</p>
<p>Would appear that a technique called “KNN” might be a good algorythm to start with. Gonna follow the tutorial the previous link gives, using the “iris” dataset</p>
<p>First steps are too really undersand your data, and too visualise your data. So:</p>
<ul>
<li><p>Sepals are modified leaves which encase and protect the flower before it blooms</p></li>
<li><p>Petals are modified leaves that surround the reproductive parts of the flower.</p></li>
</ul>
<div class="sourceCode" id="cb68"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb68-1"><a href="new-skills.html#cb68-1" aria-hidden="true" tabindex="-1"></a>iris <span class="sc">%&gt;%</span> <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>Sepal.Length, <span class="at">y=</span>Sepal.Width, <span class="at">colour=</span>Species))<span class="sc">+</span></span>
<span id="cb68-2"><a href="new-skills.html#cb68-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()</span>
<span id="cb68-3"><a href="new-skills.html#cb68-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-4"><a href="new-skills.html#cb68-4" aria-hidden="true" tabindex="-1"></a>iris <span class="sc">%&gt;%</span> <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>Petal.Length, <span class="at">y=</span>Petal.Width, <span class="at">colour=</span>Species))<span class="sc">+</span></span>
<span id="cb68-5"><a href="new-skills.html#cb68-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()</span>
<span id="cb68-6"><a href="new-skills.html#cb68-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-7"><a href="new-skills.html#cb68-7" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(iris)</span>
<span id="cb68-8"><a href="new-skills.html#cb68-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-9"><a href="new-skills.html#cb68-9" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(iris<span class="sc">$</span>Species)</span></code></pre></div>
<p>Apparently, normalising is very important in machine learning. Using summary, we can check if datasets are too far appart</p>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb69-1"><a href="new-skills.html#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(iris)</span></code></pre></div>
<p>Ranges appear to be within 0.1-7.9, a factor 10 difference. Seems reasonable enough.</p>
<p>Just in case though, you usually have to create normalising functions yourself, so here’s one anyway, created by the site and broken down for me to understand it</p>
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb70-1"><a href="new-skills.html#cb70-1" aria-hidden="true" tabindex="-1"></a>normalize<span class="ot">&lt;-</span><span class="cf">function</span>(x){</span>
<span id="cb70-2"><a href="new-skills.html#cb70-2" aria-hidden="true" tabindex="-1"></a>  num<span class="ot">&lt;-</span>x<span class="sc">-</span><span class="fu">min</span>(x)</span>
<span id="cb70-3"><a href="new-skills.html#cb70-3" aria-hidden="true" tabindex="-1"></a>  denom<span class="ot">&lt;-</span><span class="fu">max</span>(x)<span class="sc">-</span><span class="fu">min</span>(x)</span>
<span id="cb70-4"><a href="new-skills.html#cb70-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(num<span class="sc">/</span>denom)</span>
<span id="cb70-5"><a href="new-skills.html#cb70-5" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb70-6"><a href="new-skills.html#cb70-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-7"><a href="new-skills.html#cb70-7" aria-hidden="true" tabindex="-1"></a><span class="fu">normalize</span>(iris[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>]) <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span>
<span id="cb70-8"><a href="new-skills.html#cb70-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-9"><a href="new-skills.html#cb70-9" aria-hidden="true" tabindex="-1"></a><span class="co">#Generating a set where the the data is compensated for the lowest datapoint</span></span>
<span id="cb70-10"><a href="new-skills.html#cb70-10" aria-hidden="true" tabindex="-1"></a>num<span class="ot">&lt;-</span>iris<span class="sc">$</span>Sepal.Length<span class="sc">-</span><span class="fu">min</span>(iris<span class="sc">$</span>Sepal.Length)</span>
<span id="cb70-11"><a href="new-skills.html#cb70-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-12"><a href="new-skills.html#cb70-12" aria-hidden="true" tabindex="-1"></a><span class="co">#Generating the biggest difference in the dataset</span></span>
<span id="cb70-13"><a href="new-skills.html#cb70-13" aria-hidden="true" tabindex="-1"></a>denom<span class="ot">&lt;-</span><span class="fu">max</span>(iris<span class="sc">$</span>Sepal.Length)<span class="sc">-</span><span class="fu">min</span>(iris<span class="sc">$</span>Sepal.Length)</span>
<span id="cb70-14"><a href="new-skills.html#cb70-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-15"><a href="new-skills.html#cb70-15" aria-hidden="true" tabindex="-1"></a><span class="co">#Generating the size of a datapoint relative to the maximum size of a datapoint.</span></span>
<span id="cb70-16"><a href="new-skills.html#cb70-16" aria-hidden="true" tabindex="-1"></a>num<span class="sc">/</span>denom <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span>
<span id="cb70-17"><a href="new-skills.html#cb70-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-18"><a href="new-skills.html#cb70-18" aria-hidden="true" tabindex="-1"></a><span class="fu">as.data.frame</span>(<span class="fu">lapply</span>(iris[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>], normalize)) <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">Species=</span>iris<span class="sc">$</span>Species) <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code></pre></div>
<p>In order to properly train the algorythm, we must separate our data into 2 groups: a “training” group and a “test” group. Apparently, usually the “training group” is 2/3rds of the dataset, while the “test group” is one third of the dataset. Important is to give the algorythm equal amounts of each test condition, if we’d split the iris set into 2/3rds and 1/3rds it’d give 50 setosa, 50 versicolor and 0 virginica which would make the algorythm not recognize virginica at all. Thus, we use setseed and sample to generate a sample.</p>
<div class="sourceCode" id="cb71"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb71-1"><a href="new-skills.html#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1234</span>) <span class="co">#Set seed to &quot;determine&quot; the randomness&quot;</span></span>
<span id="cb71-2"><a href="new-skills.html#cb71-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-3"><a href="new-skills.html#cb71-3" aria-hidden="true" tabindex="-1"></a>ind<span class="ot">&lt;-</span><span class="fu">sample</span>(<span class="dv">2</span>, <span class="fu">nrow</span>(iris), <span class="at">replace=</span><span class="cn">TRUE</span>, <span class="at">prob=</span><span class="fu">c</span>(<span class="fl">0.67</span>, <span class="fl">0.33</span>))</span>
<span id="cb71-4"><a href="new-skills.html#cb71-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-5"><a href="new-skills.html#cb71-5" aria-hidden="true" tabindex="-1"></a>iris_training<span class="ot">&lt;-</span>iris[ind<span class="sc">==</span><span class="dv">1</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>] <span class="co">#Storing the training dataset</span></span>
<span id="cb71-6"><a href="new-skills.html#cb71-6" aria-hidden="true" tabindex="-1"></a>iris_testing<span class="ot">&lt;-</span>iris[ind<span class="sc">==</span><span class="dv">2</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>] <span class="co">#Storing the testing dataset</span></span>
<span id="cb71-7"><a href="new-skills.html#cb71-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-8"><a href="new-skills.html#cb71-8" aria-hidden="true" tabindex="-1"></a><span class="co"># We have the training and testing dataset, but we don&#39;t have the labels associated to those datasets. We have to store these seperately:</span></span>
<span id="cb71-9"><a href="new-skills.html#cb71-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-10"><a href="new-skills.html#cb71-10" aria-hidden="true" tabindex="-1"></a>iris_trainlabels<span class="ot">&lt;-</span>iris[ind<span class="sc">==</span><span class="dv">1</span>,<span class="dv">5</span>]</span>
<span id="cb71-11"><a href="new-skills.html#cb71-11" aria-hidden="true" tabindex="-1"></a>iris_testlabels<span class="ot">&lt;-</span>iris[ind<span class="sc">==</span><span class="dv">2</span>,<span class="dv">5</span>]</span></code></pre></div>
<p>With this, everything is set up to use knn: k-nearest neighbour classification.</p>
<div class="sourceCode" id="cb72"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb72-1"><a href="new-skills.html#cb72-1" aria-hidden="true" tabindex="-1"></a>iris_pred<span class="ot">&lt;-</span>class<span class="sc">::</span><span class="fu">knn</span>(<span class="at">train =</span> iris_training, <span class="at">test =</span> iris_testing, <span class="at">cl =</span> iris_trainlabels)</span>
<span id="cb72-2"><a href="new-skills.html#cb72-2" aria-hidden="true" tabindex="-1"></a>iris_testlabels</span>
<span id="cb72-3"><a href="new-skills.html#cb72-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-4"><a href="new-skills.html#cb72-4" aria-hidden="true" tabindex="-1"></a>store<span class="ot">&lt;-</span>iris_pred <span class="sc">==</span> iris_testlabels</span>
<span id="cb72-5"><a href="new-skills.html#cb72-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-6"><a href="new-skills.html#cb72-6" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(store)</span>
<span id="cb72-7"><a href="new-skills.html#cb72-7" aria-hidden="true" tabindex="-1"></a><span class="co">#Knn predicted 44 correct, 1 wrong</span></span></code></pre></div>
<p>In order to further study the model’s performance, we’ll install the package gmodels</p>
<div class="sourceCode" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb73-1"><a href="new-skills.html#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;gmodels&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="new-skills.html#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gmodels)</span>
<span id="cb74-2"><a href="new-skills.html#cb74-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb74-3"><a href="new-skills.html#cb74-3" aria-hidden="true" tabindex="-1"></a><span class="fu">CrossTable</span>(<span class="at">x =</span> iris_testlabels, <span class="at">y=</span>iris_pred)</span></code></pre></div>
<p>There were 14 setosa’s, 11 versicolors and 20 virginica in the training set (See the rows). The program identified 14 setosa’s 10 versicolors and 21 virginica’s (see the columns). 1 versicolor was thus incorrectly identified as a virginica. This is unfortunate, but significant enough to discredit this method.</p>
<p>In order to test if I’ve truly understood these past lessons, I will now perform them on a dataset of my own. For this, I’ll be using the “Glass” dataset from mlbench. This dataset contains 213 observations of a chemical analysis of 7 different types of glass.</p>
<div class="sourceCode" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb75-1"><a href="new-skills.html#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;mlbench&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb76"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb76-1"><a href="new-skills.html#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mlbench)</span>
<span id="cb76-2"><a href="new-skills.html#cb76-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(Glass)</span></code></pre></div>
<p>Firstly, we’ll take a look at the data to determine whether it needs to be normalized or not.</p>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb77-1"><a href="new-skills.html#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(Glass)</span></code></pre></div>
<p>Based on the summary data, we can tell that the biggest scale of difference between data is 0.1-75. This is a factor 100 difference, which seems pretty big. For now, we’ll just accept this and continue on with the machine learning.</p>
<p>We’ll now split out data into a train-group and a test group using the same technique as before</p>
<div class="sourceCode" id="cb78"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb78-1"><a href="new-skills.html#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">4321</span>)</span>
<span id="cb78-2"><a href="new-skills.html#cb78-2" aria-hidden="true" tabindex="-1"></a>ind_glass<span class="ot">&lt;-</span><span class="fu">sample</span>(<span class="dv">2</span>, <span class="fu">nrow</span>(Glass), <span class="at">replace=</span><span class="cn">TRUE</span>, <span class="at">prob=</span><span class="fu">c</span>(<span class="fl">0.67</span>,<span class="fl">0.33</span>)) <span class="co">#Creating a random selection of datapoints</span></span>
<span id="cb78-3"><a href="new-skills.html#cb78-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb78-4"><a href="new-skills.html#cb78-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb78-5"><a href="new-skills.html#cb78-5" aria-hidden="true" tabindex="-1"></a>glass_training<span class="ot">&lt;-</span>Glass[ind_glass<span class="sc">==</span><span class="dv">1</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">9</span>] <span class="co"># Separating the training the training and the testing datasets</span></span>
<span id="cb78-6"><a href="new-skills.html#cb78-6" aria-hidden="true" tabindex="-1"></a>glass_test<span class="ot">&lt;-</span>Glass[ind_glass<span class="sc">==</span><span class="dv">2</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">9</span>]</span>
<span id="cb78-7"><a href="new-skills.html#cb78-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb78-8"><a href="new-skills.html#cb78-8" aria-hidden="true" tabindex="-1"></a>glass_training_labels<span class="ot">&lt;-</span>Glass[ind_glass<span class="sc">==</span><span class="dv">1</span>,<span class="dv">10</span>] <span class="co"># Storing the labels seperately</span></span>
<span id="cb78-9"><a href="new-skills.html#cb78-9" aria-hidden="true" tabindex="-1"></a>glass_test_labels<span class="ot">&lt;-</span>Glass[ind_glass<span class="sc">==</span><span class="dv">2</span>,<span class="dv">10</span>]</span></code></pre></div>
<p>Now, with all datasets ready, we’ll perform the learning</p>
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb79-1"><a href="new-skills.html#cb79-1" aria-hidden="true" tabindex="-1"></a>glass_pred<span class="ot">&lt;-</span>class<span class="sc">::</span><span class="fu">knn</span>(glass_training, glass_test, glass_training_labels) <span class="co">#Performing the machine learning test.</span></span>
<span id="cb79-2"><a href="new-skills.html#cb79-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-3"><a href="new-skills.html#cb79-3" aria-hidden="true" tabindex="-1"></a>glass_result<span class="ot">&lt;-</span>glass_pred <span class="sc">==</span> glass_test_labels</span>
<span id="cb79-4"><a href="new-skills.html#cb79-4" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(glass_result)</span>
<span id="cb79-5"><a href="new-skills.html#cb79-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-6"><a href="new-skills.html#cb79-6" aria-hidden="true" tabindex="-1"></a><span class="fu">CrossTable</span>(<span class="at">x =</span> glass_test_labels, <span class="at">y=</span>glass_pred)</span></code></pre></div>
<p>For this dataset, KNN was less accurate: from the 70 test subjects, it managed to identify 55 correctly, but miss-identified 15 of them. As such, we can immidiately see that KNN is not perfect. However, this <em>does</em> prove that I can now use a basic machine learning algorythm in R.</p>
<p>Just to sate my curiousity, we’ll now look at the results which we would’ve gotten if we’d normalised the data.</p>
<div class="sourceCode" id="cb80"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb80-1"><a href="new-skills.html#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">4321</span>)</span>
<span id="cb80-2"><a href="new-skills.html#cb80-2" aria-hidden="true" tabindex="-1"></a>ind_glass<span class="ot">&lt;-</span><span class="fu">sample</span>(<span class="dv">2</span>, <span class="fu">nrow</span>(Glass), <span class="at">replace=</span><span class="cn">TRUE</span>, <span class="at">prob=</span><span class="fu">c</span>(<span class="fl">0.67</span>,<span class="fl">0.33</span>)) <span class="co">#Creating a random selection of datapoints</span></span>
<span id="cb80-3"><a href="new-skills.html#cb80-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-4"><a href="new-skills.html#cb80-4" aria-hidden="true" tabindex="-1"></a>Glass_norm<span class="ot">&lt;-</span><span class="fu">normalize</span>(Glass[<span class="dv">1</span><span class="sc">:</span><span class="dv">9</span>]) <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">Type=</span>Glass<span class="sc">$</span>Type)</span>
<span id="cb80-5"><a href="new-skills.html#cb80-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-6"><a href="new-skills.html#cb80-6" aria-hidden="true" tabindex="-1"></a>glass_training<span class="ot">&lt;-</span>Glass_norm[ind_glass<span class="sc">==</span><span class="dv">1</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">9</span>] <span class="co"># Separating the training the training and the testing datasets</span></span>
<span id="cb80-7"><a href="new-skills.html#cb80-7" aria-hidden="true" tabindex="-1"></a>glass_test<span class="ot">&lt;-</span>Glass_norm[ind_glass<span class="sc">==</span><span class="dv">2</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">9</span>]</span>
<span id="cb80-8"><a href="new-skills.html#cb80-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-9"><a href="new-skills.html#cb80-9" aria-hidden="true" tabindex="-1"></a>glass_training_labels<span class="ot">&lt;-</span>Glass_norm[ind_glass<span class="sc">==</span><span class="dv">1</span>,<span class="dv">10</span>] <span class="co"># Storing the labels seperately</span></span>
<span id="cb80-10"><a href="new-skills.html#cb80-10" aria-hidden="true" tabindex="-1"></a>glass_test_labels<span class="ot">&lt;-</span>Glass_norm[ind_glass<span class="sc">==</span><span class="dv">2</span>,<span class="dv">10</span>]</span>
<span id="cb80-11"><a href="new-skills.html#cb80-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-12"><a href="new-skills.html#cb80-12" aria-hidden="true" tabindex="-1"></a>glass_pred<span class="ot">&lt;-</span>class<span class="sc">::</span><span class="fu">knn</span>(glass_training, glass_test, glass_training_labels) <span class="co">#Performing the machine learning test.</span></span>
<span id="cb80-13"><a href="new-skills.html#cb80-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-14"><a href="new-skills.html#cb80-14" aria-hidden="true" tabindex="-1"></a>glass_result<span class="ot">&lt;-</span>glass_pred <span class="sc">==</span> glass_test_labels</span>
<span id="cb80-15"><a href="new-skills.html#cb80-15" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(glass_result)</span>
<span id="cb80-16"><a href="new-skills.html#cb80-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-17"><a href="new-skills.html#cb80-17" aria-hidden="true" tabindex="-1"></a><span class="fu">CrossTable</span>(<span class="at">x =</span> glass_test_labels, <span class="at">y=</span>glass_pred)</span></code></pre></div>
<p>The normalised test gives the exact same result as the non-normalised test.</p>
<p>Now, as a small sidetrack, we’ll practice with the “karet” package, “classification and regression training”. This is a more generalised way to do different kinds of machine learning</p>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb81-1"><a href="new-skills.html#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;caret&quot;</span>)</span></code></pre></div>
<p>This time, the data will be split with a 75-25 ratio</p>
<div class="sourceCode" id="cb82"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb82-1"><a href="new-skills.html#cb82-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(caret)</span>
<span id="cb82-2"><a href="new-skills.html#cb82-2" aria-hidden="true" tabindex="-1"></a>index<span class="ot">&lt;-</span><span class="fu">createDataPartition</span>(iris<span class="sc">$</span>Species, <span class="at">p=</span><span class="fl">0.75</span>, <span class="at">list=</span><span class="cn">FALSE</span>)</span>
<span id="cb82-3"><a href="new-skills.html#cb82-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-4"><a href="new-skills.html#cb82-4" aria-hidden="true" tabindex="-1"></a>iris_training2<span class="ot">&lt;-</span>iris[index,]</span>
<span id="cb82-5"><a href="new-skills.html#cb82-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-6"><a href="new-skills.html#cb82-6" aria-hidden="true" tabindex="-1"></a>iris_test2<span class="ot">&lt;-</span>iris[<span class="sc">-</span>index,]</span></code></pre></div>
<p>There’s a <em>lot</em> of algorythms in Caret, so we’ll use getModelInfo() to get info about these models. There’s over 230 models, so only the top 50 will be shown</p>
<div class="sourceCode" id="cb83"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb83-1"><a href="new-skills.html#cb83-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(<span class="fu">getModelInfo</span>()) <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="dv">50</span>)</span>
<span id="cb83-2"><a href="new-skills.html#cb83-2" aria-hidden="true" tabindex="-1"></a><span class="co">#We&#39;ll once again just use knn. But now</span></span>
<span id="cb83-3"><a href="new-skills.html#cb83-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-4"><a href="new-skills.html#cb83-4" aria-hidden="true" tabindex="-1"></a><span class="co">#First, we create and train the algorythm</span></span>
<span id="cb83-5"><a href="new-skills.html#cb83-5" aria-hidden="true" tabindex="-1"></a>model_knn<span class="ot">&lt;-</span><span class="fu">train</span>(iris_training2[,<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>], iris_training2[,<span class="dv">5</span>], <span class="at">method=</span><span class="st">&#39;knn&#39;</span>)</span>
<span id="cb83-6"><a href="new-skills.html#cb83-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-7"><a href="new-skills.html#cb83-7" aria-hidden="true" tabindex="-1"></a><span class="co">#Then, we use the model to make a prediction</span></span>
<span id="cb83-8"><a href="new-skills.html#cb83-8" aria-hidden="true" tabindex="-1"></a>prediction<span class="ot">&lt;-</span><span class="fu">predict</span>(<span class="at">object=</span>model_knn, iris_test2[,<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>])</span>
<span id="cb83-9"><a href="new-skills.html#cb83-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-10"><a href="new-skills.html#cb83-10" aria-hidden="true" tabindex="-1"></a><span class="co">#We check if the predicitons are true</span></span>
<span id="cb83-11"><a href="new-skills.html#cb83-11" aria-hidden="true" tabindex="-1"></a>pred<span class="ot">&lt;-</span>prediction <span class="sc">==</span> iris_test2[,<span class="dv">5</span>]</span>
<span id="cb83-12"><a href="new-skills.html#cb83-12" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(pred)</span>
<span id="cb83-13"><a href="new-skills.html#cb83-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-14"><a href="new-skills.html#cb83-14" aria-hidden="true" tabindex="-1"></a><span class="co">#And we create a confusion matrix for the results</span></span>
<span id="cb83-15"><a href="new-skills.html#cb83-15" aria-hidden="true" tabindex="-1"></a><span class="fu">confusionMatrix</span>(prediction,iris_test2[,<span class="dv">5</span>])</span></code></pre></div>
<p>In the caret package, there’s also built-in pre-processing options, as shown below</p>
<div class="sourceCode" id="cb84"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb84-1"><a href="new-skills.html#cb84-1" aria-hidden="true" tabindex="-1"></a>model_knn<span class="ot">&lt;-</span><span class="fu">train</span>(iris_training2[,<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>],iris_training2[,<span class="dv">5</span>],<span class="at">method=</span><span class="st">&#39;knn&#39;</span>,<span class="at">preProcess =</span> <span class="fu">c</span>(<span class="st">&quot;center&quot;</span>, <span class="st">&quot;scale&quot;</span>))</span>
<span id="cb84-2"><a href="new-skills.html#cb84-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-3"><a href="new-skills.html#cb84-3" aria-hidden="true" tabindex="-1"></a>prediction<span class="ot">&lt;-</span><span class="fu">predict</span>(<span class="at">object=</span>model_knn, iris_test2[,<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>], <span class="at">type=</span><span class="st">&#39;raw&#39;</span>)</span>
<span id="cb84-4"><a href="new-skills.html#cb84-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-5"><a href="new-skills.html#cb84-5" aria-hidden="true" tabindex="-1"></a><span class="fu">confusionMatrix</span>(prediction, iris_test2[,<span class="dv">5</span>])</span></code></pre></div>
<p>Using these forms of pr-processing give us 1 less mistake, and thus a more accurate system.</p>
<p>Once again, to properly demonstrate my ability in now using the carets package, I’ll be using it on the glass dataset once again</p>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb85-1"><a href="new-skills.html#cb85-1" aria-hidden="true" tabindex="-1"></a>Glass_part<span class="ot">&lt;-</span><span class="fu">createDataPartition</span>(Glass<span class="sc">$</span>Type, <span class="at">p=</span><span class="fl">0.75</span>, <span class="at">list=</span><span class="cn">FALSE</span>)</span>
<span id="cb85-2"><a href="new-skills.html#cb85-2" aria-hidden="true" tabindex="-1"></a>Glass_training2<span class="ot">&lt;-</span>Glass[Glass_part,]</span>
<span id="cb85-3"><a href="new-skills.html#cb85-3" aria-hidden="true" tabindex="-1"></a>Glass_test2<span class="ot">&lt;-</span>Glass[<span class="sc">-</span>Glass_part,]</span>
<span id="cb85-4"><a href="new-skills.html#cb85-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb85-5"><a href="new-skills.html#cb85-5" aria-hidden="true" tabindex="-1"></a>model_knn2<span class="ot">&lt;-</span><span class="fu">train</span>(Glass_training2[,<span class="dv">1</span><span class="sc">:</span><span class="dv">9</span>], Glass_training2[,<span class="dv">10</span>], <span class="at">method =</span> <span class="st">&quot;knn&quot;</span>)</span>
<span id="cb85-6"><a href="new-skills.html#cb85-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb85-7"><a href="new-skills.html#cb85-7" aria-hidden="true" tabindex="-1"></a>Prediction<span class="ot">&lt;-</span><span class="fu">predict</span>(<span class="at">object=</span>model_knn2, Glass_test2)</span>
<span id="cb85-8"><a href="new-skills.html#cb85-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb85-9"><a href="new-skills.html#cb85-9" aria-hidden="true" tabindex="-1"></a><span class="fu">confusionMatrix</span>(Prediction, Glass_test2[,<span class="dv">10</span>])</span></code></pre></div>
<p>Without pre-processing nor setting the data raw, the prediction comes down to a accuracy of .69</p>
<div class="sourceCode" id="cb86"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb86-1"><a href="new-skills.html#cb86-1" aria-hidden="true" tabindex="-1"></a>Glass_part<span class="ot">&lt;-</span><span class="fu">createDataPartition</span>(Glass<span class="sc">$</span>Type, <span class="at">p=</span><span class="fl">0.75</span>, <span class="at">list=</span><span class="cn">FALSE</span>)</span>
<span id="cb86-2"><a href="new-skills.html#cb86-2" aria-hidden="true" tabindex="-1"></a>Glass_training2<span class="ot">&lt;-</span>Glass[Glass_part,]</span>
<span id="cb86-3"><a href="new-skills.html#cb86-3" aria-hidden="true" tabindex="-1"></a>Glass_test2<span class="ot">&lt;-</span>Glass[<span class="sc">-</span>Glass_part,]</span>
<span id="cb86-4"><a href="new-skills.html#cb86-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-5"><a href="new-skills.html#cb86-5" aria-hidden="true" tabindex="-1"></a>model_knn3<span class="ot">&lt;-</span><span class="fu">train</span>(Glass_training2[,<span class="dv">1</span><span class="sc">:</span><span class="dv">9</span>], Glass_training2[,<span class="dv">10</span>], <span class="at">method =</span> <span class="st">&quot;knn&quot;</span>, <span class="at">preProcess =</span> <span class="fu">c</span>(<span class="st">&quot;center&quot;</span>, <span class="st">&quot;scale&quot;</span>))</span>
<span id="cb86-6"><a href="new-skills.html#cb86-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-7"><a href="new-skills.html#cb86-7" aria-hidden="true" tabindex="-1"></a>Prediction<span class="ot">&lt;-</span><span class="fu">predict</span>(<span class="at">object=</span>model_knn3, Glass_test2, <span class="at">type =</span> <span class="st">&quot;raw&quot;</span>)</span>
<span id="cb86-8"><a href="new-skills.html#cb86-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-9"><a href="new-skills.html#cb86-9" aria-hidden="true" tabindex="-1"></a><span class="fu">confusionMatrix</span>(Prediction, Glass_test2[,<span class="dv">10</span>])</span></code></pre></div>
<p>Using pre-processing, the accuracy <em>decreased</em> by 0.04, to a accuracy of .65</p>
<p>Thus, we can conclude: pre-processing is <em>not</em> always beneficial if done without knowledge of what the pre-processing does.</p>
<p>Now that we understand the <em>basics</em> of machine learning, we’ll try to go a little deeper in the specific subject I want: machine learning in microbiology. For this, we’ll study <a href="https://www.frontiersin.org/articles/10.3389/fmicb.2019.00827/full">this article</a>. Underneath are the notes I’ve taken while reading through this article:</p>
<ul>
<li><p>Machine learning has two forms: supervised and unsupervised learning. Supervised requires a specific training package, unsupervised uses clustering, adapts k-means to reduce erorr through iteration.</p></li>
<li><p>Machine learning in microbiology has 4 steps: Extraction of features, Operational classifaction units (OTU) by clustering, important features selection to improve accuracy, training dataset to train the model -&gt; testing dataset to test the model.</p></li>
<li><p>Raw microbe data has high data dimensions: the data is too complex. Thus, the data’s complexity needs to be reduced. For this, usually a PCA / PCoA analysis is usually performed.</p></li>
<li><p>In microbial study, supervised studying is usually used: using support vector machine (SVM), Naïve Bayes (NB), random forest (RF) or K nearest neighbour (KNN) methods.</p></li>
<li><p>In machine learning, there’s two different types of identification: determining whether an organism belongs to a species or not (is it streptococcus pneumoniae or is it <em>something else</em>), or classification based on domain: which kingdom, phylum, class, etc. does the micro-organism belong too?</p></li>
<li><p>Next too identification of microbes, machine learning can also be used to predict of enviromental and host phenotypes or too predict disease using microbial communities.</p></li>
<li><p>Finally, machine learning can also be used to look at interactions between micro-organisms and microbiome-disease associations.</p></li>
</ul>
<p>Now, to apply all this knowledge gained from reading the article, we’ll look for a referenced article which has their data and code open sourced, and then use that to further figure out how to apply machine learning in practical applications. For this, we’ll use the IDTAXA function <a href="https://www.frontiersin.org/articles/10.3389/fgene.2018.00304/full">from this article</a></p>
<div class="sourceCode" id="cb87"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb87-1"><a href="new-skills.html#cb87-1" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">&quot;DECIPHER&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb88-1"><a href="new-skills.html#cb88-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(BiocManager)</span>
<span id="cb88-2"><a href="new-skills.html#cb88-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DECIPHER)</span>
<span id="cb88-3"><a href="new-skills.html#cb88-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb88-4"><a href="new-skills.html#cb88-4" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;TrainingSet_16S&quot;</span>)</span>
<span id="cb88-5"><a href="new-skills.html#cb88-5" aria-hidden="true" tabindex="-1"></a>TrainingSet_16S</span></code></pre></div>
<p>By looking around at some of the data a functions. Under the <a href="http://www2.decipher.codes/Gallery.html">DECIPHER site</a> we can find a tutorial on how “DECIPHER” can be used. To test the functions out, I’ve taken 3 random bacterial DNA sequences: One from <a href="https://www.ncbi.nlm.nih.gov/genome/175?genome_assembly_id=383877">streptococcus pyogenes</a>, one from <a href="https://www.ncbi.nlm.nih.gov/genome/154?genome_assembly_id=360991">staphylococcus aureus</a> and finally, one from <a href="https://www.ncbi.nlm.nih.gov/genome/152?genome_assembly_id=299225">Salmonella enterica typhi</a>. Before that, however, we’ll use the built-in example given by DECIPHER to truly study how decipher works.</p>
<div class="sourceCode" id="cb89"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb89-1"><a href="new-skills.html#cb89-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Loading sequences into database</span></span>
<span id="cb89-2"><a href="new-skills.html#cb89-2" aria-hidden="true" tabindex="-1"></a>fas<span class="ot">&lt;-</span><span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;50S_ribosomal_protein_L2.fas&quot;</span>, <span class="at">package=</span><span class="st">&quot;DECIPHER&quot;</span>) <span class="co">#First we take a built-in data set</span></span>
<span id="cb89-3"><a href="new-skills.html#cb89-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-4"><a href="new-skills.html#cb89-4" aria-hidden="true" tabindex="-1"></a>dbConn<span class="ot">&lt;-</span><span class="fu">dbConnect</span>(<span class="fu">SQLite</span>(),<span class="st">&quot;:memory:&quot;</span>) <span class="co">#Then we set up a connection to a database</span></span>
<span id="cb89-5"><a href="new-skills.html#cb89-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-6"><a href="new-skills.html#cb89-6" aria-hidden="true" tabindex="-1"></a><span class="fu">Seqs2DB</span>(fas, <span class="at">type=</span><span class="st">&quot;FASTA&quot;</span>, <span class="at">dbFile =</span> dbConn, <span class="at">identifier=</span><span class="st">&quot;&quot;</span>) <span class="co">#Now we&#39;re adding out fasta data to a database from the file</span></span>
<span id="cb89-7"><a href="new-skills.html#cb89-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-8"><a href="new-skills.html#cb89-8" aria-hidden="true" tabindex="-1"></a><span class="co">#Identifiying sequences based on genus</span></span>
<span id="cb89-9"><a href="new-skills.html#cb89-9" aria-hidden="true" tabindex="-1"></a>x<span class="ot">&lt;-</span><span class="fu">dbGetQuery</span>(dbConn, <span class="st">&quot;Select description from Seqs&quot;</span>)<span class="sc">$</span>description <span class="co">#Getting the description values out of the database</span></span>
<span id="cb89-10"><a href="new-skills.html#cb89-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-11"><a href="new-skills.html#cb89-11" aria-hidden="true" tabindex="-1"></a>ns<span class="ot">&lt;-</span><span class="fu">unlist</span>(<span class="fu">lapply</span>(<span class="fu">strsplit</span>(x, <span class="at">split=</span><span class="st">&quot; &quot;</span>), <span class="co">#I have no idea _how_ it works, but this extracts the genus names from all bacteria</span></span>
<span id="cb89-12"><a href="new-skills.html#cb89-12" aria-hidden="true" tabindex="-1"></a>                  <span class="at">FUN=</span><span class="st">`</span><span class="at">[</span><span class="st">`</span>,1L))</span>
<span id="cb89-13"><a href="new-skills.html#cb89-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-14"><a href="new-skills.html#cb89-14" aria-hidden="true" tabindex="-1"></a><span class="fu">Add2DB</span>(<span class="at">myData =</span> <span class="fu">data.frame</span>(<span class="at">identifier=</span>ns, <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>), <span class="at">dbFile =</span> dbConn) <span class="co">#Add</span></span>
<span id="cb89-15"><a href="new-skills.html#cb89-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-16"><a href="new-skills.html#cb89-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Align the translated sequences</span></span>
<span id="cb89-17"><a href="new-skills.html#cb89-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-18"><a href="new-skills.html#cb89-18" aria-hidden="true" tabindex="-1"></a>dna<span class="ot">&lt;-</span><span class="fu">SearchDB</span>(dbConn, <span class="at">nameBy =</span> <span class="st">&quot;identifier&quot;</span>)</span>
<span id="cb89-19"><a href="new-skills.html#cb89-19" aria-hidden="true" tabindex="-1"></a>AA<span class="ot">&lt;-</span><span class="fu">AlignTranslation</span>(dna, <span class="at">type=</span><span class="st">&quot;AAStringSet&quot;</span>)</span>
<span id="cb89-20"><a href="new-skills.html#cb89-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-21"><a href="new-skills.html#cb89-21" aria-hidden="true" tabindex="-1"></a><span class="fu">Seqs2DB</span>(AA, <span class="at">type =</span> <span class="st">&quot;AAStringSet&quot;</span>,</span>
<span id="cb89-22"><a href="new-skills.html#cb89-22" aria-hidden="true" tabindex="-1"></a>        <span class="at">dbFile =</span> dbConn,</span>
<span id="cb89-23"><a href="new-skills.html#cb89-23" aria-hidden="true" tabindex="-1"></a>        <span class="at">identifier =</span> <span class="st">&quot;&quot;</span>,</span>
<span id="cb89-24"><a href="new-skills.html#cb89-24" aria-hidden="true" tabindex="-1"></a>        <span class="at">tblName =</span> <span class="st">&quot;Aligned&quot;</span>)</span>
<span id="cb89-25"><a href="new-skills.html#cb89-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-26"><a href="new-skills.html#cb89-26" aria-hidden="true" tabindex="-1"></a><span class="fu">Add2DB</span>(<span class="at">myData =</span> <span class="fu">data.frame</span>(<span class="at">identifier=</span>ns, <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span>),</span>
<span id="cb89-27"><a href="new-skills.html#cb89-27" aria-hidden="true" tabindex="-1"></a>       <span class="at">dbFile =</span> dbConn,</span>
<span id="cb89-28"><a href="new-skills.html#cb89-28" aria-hidden="true" tabindex="-1"></a>       <span class="at">tblName =</span> <span class="st">&quot;Aligned&quot;</span>)</span>
<span id="cb89-29"><a href="new-skills.html#cb89-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-30"><a href="new-skills.html#cb89-30" aria-hidden="true" tabindex="-1"></a><span class="co"># Form a concensus sequence for each genus</span></span>
<span id="cb89-31"><a href="new-skills.html#cb89-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-32"><a href="new-skills.html#cb89-32" aria-hidden="true" tabindex="-1"></a>cons<span class="ot">&lt;-</span><span class="fu">IdConsensus</span>(dbConn, </span>
<span id="cb89-33"><a href="new-skills.html#cb89-33" aria-hidden="true" tabindex="-1"></a>            <span class="at">tblName =</span> <span class="st">&quot;Aligned&quot;</span>,</span>
<span id="cb89-34"><a href="new-skills.html#cb89-34" aria-hidden="true" tabindex="-1"></a>            <span class="at">type=</span><span class="st">&quot;AAStringSet&quot;</span>,</span>
<span id="cb89-35"><a href="new-skills.html#cb89-35" aria-hidden="true" tabindex="-1"></a>            <span class="at">threshold=</span><span class="fl">0.5</span>,</span>
<span id="cb89-36"><a href="new-skills.html#cb89-36" aria-hidden="true" tabindex="-1"></a>            <span class="at">minInformation=</span><span class="fl">0.5</span>)</span>
<span id="cb89-37"><a href="new-skills.html#cb89-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-38"><a href="new-skills.html#cb89-38" aria-hidden="true" tabindex="-1"></a><span class="fu">Seqs2DB</span>(cons, <span class="at">type =</span> <span class="st">&quot;DNAStringSet&quot;</span>, <span class="at">dbFile =</span> dbConn, <span class="at">identifier =</span> <span class="st">&quot;&quot;</span>, <span class="at">tblName =</span> <span class="st">&quot;Consensus&quot;</span>)</span>
<span id="cb89-39"><a href="new-skills.html#cb89-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-40"><a href="new-skills.html#cb89-40" aria-hidden="true" tabindex="-1"></a>aa<span class="ot">&lt;-</span><span class="fu">SearchDB</span>(dbConn, <span class="st">&quot;Consensus&quot;</span>,<span class="at">type =</span> <span class="st">&quot;AAStringSet&quot;</span>, <span class="at">nameBy =</span> <span class="st">&quot;description&quot;</span>, <span class="at">limit=</span><span class="st">&quot;40,20&quot;</span>, <span class="at">removeGaps =</span> <span class="st">&quot;common&quot;</span>)</span>
<span id="cb89-41"><a href="new-skills.html#cb89-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-42"><a href="new-skills.html#cb89-42" aria-hidden="true" tabindex="-1"></a><span class="co"># Now see the sequences fully</span></span>
<span id="cb89-43"><a href="new-skills.html#cb89-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-44"><a href="new-skills.html#cb89-44" aria-hidden="true" tabindex="-1"></a><span class="fu">BrowseSeqs</span>(aa, <span class="at">threshold=</span><span class="fl">0.5</span>,<span class="at">minInformation=</span><span class="fl">0.5</span>,<span class="at">colWidth =</span> <span class="dv">60</span>)</span></code></pre></div>
<p>Only after performing these steps, do I realise they have nothing to do with machine learning. oops.</p>
<p>Now, with this new knowledge, we’ll focus on the ACTUAL machine learning part of DECIPHER: IDTAXA</p>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb90-1"><a href="new-skills.html#cb90-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DECIPHER)</span>
<span id="cb90-2"><a href="new-skills.html#cb90-2" aria-hidden="true" tabindex="-1"></a>fas<span class="ot">&lt;-</span><span class="st">&quot;./data.raw/sequence_salmonella.fasta&quot;</span></span>
<span id="cb90-3"><a href="new-skills.html#cb90-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-4"><a href="new-skills.html#cb90-4" aria-hidden="true" tabindex="-1"></a>seqs<span class="ot">&lt;-</span><span class="fu">readDNAStringSet</span>(fas)</span>
<span id="cb90-5"><a href="new-skills.html#cb90-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-6"><a href="new-skills.html#cb90-6" aria-hidden="true" tabindex="-1"></a>seqs<span class="ot">&lt;-</span><span class="fu">RemoveGaps</span>(seqs)</span></code></pre></div>
<p>As the data set used for the training of the algorythm is larger than 1GB, it cannot be stored on github. Thus, you’ll have to manually install it from <a href="http://www2.decipher.codes/Downloads.html">the DECIPHER downloads page</a>, under code “KEGG ALL”</p>
<div class="sourceCode" id="cb91"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb91-1"><a href="new-skills.html#cb91-1" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">&quot;../KEGG_AllLineage_r95.RData&quot;</span>)</span>
<span id="cb91-2"><a href="new-skills.html#cb91-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-3"><a href="new-skills.html#cb91-3" aria-hidden="true" tabindex="-1"></a>ids<span class="ot">&lt;-</span><span class="fu">IdTaxa</span>(seqs,</span>
<span id="cb91-4"><a href="new-skills.html#cb91-4" aria-hidden="true" tabindex="-1"></a>            trainingSet,</span>
<span id="cb91-5"><a href="new-skills.html#cb91-5" aria-hidden="true" tabindex="-1"></a>            <span class="at">strand =</span> <span class="st">&quot;both&quot;</span>,</span>
<span id="cb91-6"><a href="new-skills.html#cb91-6" aria-hidden="true" tabindex="-1"></a>            <span class="at">threshold =</span> <span class="dv">60</span>,</span>
<span id="cb91-7"><a href="new-skills.html#cb91-7" aria-hidden="true" tabindex="-1"></a>            <span class="at">processors =</span> <span class="cn">NULL</span>)</span>
<span id="cb91-8"><a href="new-skills.html#cb91-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-9"><a href="new-skills.html#cb91-9" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(ids)</span>
<span id="cb91-10"><a href="new-skills.html#cb91-10" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(ids)</span></code></pre></div>
<p>Using this training set, the machine learning algorythm could <em>not</em> figure out what salmonella was. Instead, we’ll try it using another training data set, specifically that of proteobacteria (which salmonella is). If this algorythm cannot identify salmonella using this secondd dataset, we’ll look into the FASTA input we’re using</p>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb92-1"><a href="new-skills.html#cb92-1" aria-hidden="true" tabindex="-1"></a>fas<span class="ot">&lt;-</span><span class="st">&quot;./data.raw/sequence_salmonella.fasta&quot;</span></span>
<span id="cb92-2"><a href="new-skills.html#cb92-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb92-3"><a href="new-skills.html#cb92-3" aria-hidden="true" tabindex="-1"></a>seqs<span class="ot">&lt;-</span><span class="fu">readDNAStringSet</span>(fas)</span>
<span id="cb92-4"><a href="new-skills.html#cb92-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb92-5"><a href="new-skills.html#cb92-5" aria-hidden="true" tabindex="-1"></a>seqs<span class="ot">&lt;-</span><span class="fu">RemoveGaps</span>(seqs)</span>
<span id="cb92-6"><a href="new-skills.html#cb92-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb92-7"><a href="new-skills.html#cb92-7" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">&quot;../KEGG_Proteobacteria_r95.RData&quot;</span>)</span>
<span id="cb92-8"><a href="new-skills.html#cb92-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb92-9"><a href="new-skills.html#cb92-9" aria-hidden="true" tabindex="-1"></a>ids<span class="ot">&lt;-</span><span class="fu">IdTaxa</span>(seqs,</span>
<span id="cb92-10"><a href="new-skills.html#cb92-10" aria-hidden="true" tabindex="-1"></a>            trainingSet,</span>
<span id="cb92-11"><a href="new-skills.html#cb92-11" aria-hidden="true" tabindex="-1"></a>            <span class="at">strand =</span> <span class="st">&quot;both&quot;</span>,</span>
<span id="cb92-12"><a href="new-skills.html#cb92-12" aria-hidden="true" tabindex="-1"></a>            <span class="at">threshold =</span> <span class="dv">60</span>,</span>
<span id="cb92-13"><a href="new-skills.html#cb92-13" aria-hidden="true" tabindex="-1"></a>            <span class="at">processors =</span> <span class="cn">NULL</span>)</span>
<span id="cb92-14"><a href="new-skills.html#cb92-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb92-15"><a href="new-skills.html#cb92-15" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(ids)</span></code></pre></div>
<p>ids once again does not give any classification: We’ll try a standard dataset this time in order to see if the program is working at all.</p>

</div>
<p style="text-align: center;">
<a href="future-endeavours.html"><button class="btn btn-default">Previous</button></a>
<a href="cv.html"><button class="btn btn-default">Next</button></a>
</p>
</div>
</div>


</div>

<script>

// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});

</script>

</body>
</html>
